src/cpu/o3/lsq.cc

```
@@ -230,65 +249,91 @@ LSQ::tick()
+unsigned
+LSQ::getDcacheDiv(Addr vaddr) const
+{
+    return (vaddr >> dcacheLineBits) & (dcacheSetDivNum - 1);
+}
+
+uint64_t
+LSQ::getDcacheBankSetKey(Addr vaddr) const
+{
+    // [setIndex][bankIndex][dataOffset]
+    //         ^ (cacheLineBits)   ^ (3 bits)
+    return (vaddr >> 3) & ((1ULL << dcacheSetBankBits) - 1);
```

Cppcheck (2.19.0)

[error] Shifting by a negative value is undefined behaviour

(shiftNegative)